File: sdk/web/src/full/speaker-gate.ts
// 声纹门控完整版：VAD 切段 + 声纹判定。声纹注册/比对/短注册阈值补偿全部复用
// 同包 ../core 的 VoiceFilter，本类只负责「VAD 切段 → 逐段喂给 VoiceFilter」
// 以及注册音频的静音剥离。
// 与 app 后端 speaker_gate.py 同一套模型和判定语义；区别是 v0 按段整段判定
// （app 另有段内 0.6s 预热的增量判定和接受前缀补发，属于 ASR 联动的实时性优化，
// SDK 后续版本再做）。
import { VoiceFilter, type EnrollResult } from '../core/voice-filter'
import { createSileroVad, type SileroVadOptions, type Vad } from './vad'

export const SAMPLE_RATE = 16000

export interface SpeakerGateConfig {
  /** silero_vad.onnx 路径 */
  vadModel: string
  /** 3D-Speaker ER2Net onnx 路径 */
  speakerModel: string
  /** 放行所需的相似度阈值基准，默认 0.5（与 app 后端一致） */
  threshold?: number
  /**
   * 注册的净语音目标（秒）：流式注册累计到该时长即视为足够，说完自动完成。
   * 默认 1.2 秒 ≈ 四~六个汉字的正常语速。
   */
  enrollTargetSpeechSeconds?: number
  /** 注册允许的最少净语音（秒），不足则 finishEnroll 抛错。默认 0.6 */
  enrollMinSpeechSeconds?: number
  /** 短注册阈值补偿系数（透传给 VoiceFilter），默认 0.7，设 1 禁用 */
  shortEnrollThresholdFactor?: number
  /** VAD 参数覆盖，默认同 app 后端（同时作用于主 VAD 与注册 VAD） */
  vad?: Partial<SileroVadOptions>
}

export interface GateSegmentEvent {
  type: 'segment'
  /** 段起点在整条流中的采样序号 */
  start: number
  /** 段时长（秒） */
  durationSeconds: number
  /** 与注册声纹的余弦相似度；未注册时为 null（全放行） */
  similarity: number | null
  accepted: boolean
  /** 段内音频（float32 [-1,1] @16k），宿主可自行播放/送 ASR */
  samples: Float32Array
}

/** 流式注册进度：已累计的净语音时长、是否已达目标。 */
export interface EnrollProgress {
  speechSeconds: number
  enough: boolean
}

export type { EnrollResult }

function concatFloat32(parts: Float32Array[]): Float32Array {
  const total = parts.reduce((n, p) => n + p.length, 0)
  const out = new Float32Array(total)
  let offset = 0
  for (const p of parts) {
    out.set(p, offset)
    offset += p.length
  }
  return out
}

export class SpeakerGate {
  private constructor(
    /** 放行所需的相似度阈值基准 */
    readonly threshold: number,
    private readonly vadModelPath: string,
    private readonly vadOverrides: Partial<SileroVadOptions>,
    private readonly enrollTarget: number,
    private readonly enrollMin: number,
    private readonly vad: Vad,
    private readonly filter: VoiceFilter,
    private enrollVad: Vad | null,
    private enrollChunks: Float32Array[],
    private enrollSpeechSamples: number,
  ) {}

  static async create(config: SpeakerGateConfig): Promise<SpeakerGate> {
    const vad = createSileroVad(config.vadModel, config.vad)
    const filter = await VoiceFilter.create({
      speakerModel: config.speakerModel,
      threshold: config.threshold,
      shortEnrollThresholdFactor: config.shortEnrollThresholdFactor,
    })
    return new SpeakerGate(
      config.threshold ?? 0.5,
      config.vadModel,
      config.vad ?? {},
      config.enrollTargetSpeechSeconds ?? 1.2,
      config.enrollMinSpeechSeconds ?? 0.6,
      vad,
      filter,
      null,
      [],
      0,
    )
  }

  /** 是否已有注册声纹（未注册时所有段放行）。 */
  get enrolled(): boolean {
    return this.filter.enrolled
  }

  /** 当前实际生效的放行阈值（短注册补偿后）。 */
  get effectiveThreshold(): number {
    return this.filter.effectiveThreshold
  }

  // ── 批式注册（一次性给整段音频）────────────────────────────
  /**
   * 便捷注册：整段音频内部剥静音后提声纹（静音不参与特征统计，短注册更稳）。
   * 返回实际使用的净语音时长；未检测到足够语音时抛错。
   */
  async enroll(samples: Float32Array): Promise<EnrollResult> {
    this.beginEnroll()
    await this.enrollChunk(samples)
    return this.finishEnroll()
  }

  // ── 流式注册（边说边收，够量自动完成）──────────────────────
  /** 开始一次流式注册：复位累计状态。 */
  beginEnroll(): void {
    this.enrollVad?.free()
    this.enrollVad = createSileroVad(this.vadModelPath, this.vadOverrides)
    this.enrollChunks = []
    this.enrollSpeechSamples = 0
  }

  /**
   * 喂入一段注册音频（可与 accept 并行使用，注册走独立 VAD 不影响主流）。
   * 返回当前净语音进度；enough=true 表示已达目标，宿主应停麦并 finishEnroll()。
   */
  async enrollChunk(samples: Float32Array): Promise<EnrollProgress> {
    if (!this.enrollVad) this.beginEnroll()
    this.enrollVad!.acceptWaveform(samples)
    this.collectEnrollSegments()
    const speechSeconds = this.enrollSpeechSamples / SAMPLE_RATE
    return { speechSeconds, enough: speechSeconds >= this.enrollTarget }
  }

  /**
   * 结束注册（冲出尾部语音段后）提声纹。
   * 净语音低于下限时抛错（调用方提示「没听清/再说一句」），注册状态保持未完成。
   */
  async finishEnroll(): Promise<EnrollResult> {
    if (!this.enrollVad) throw new Error('注册未开始（先 beginEnroll）')
    this.enrollVad.flush()
    this.collectEnrollSegments()
    this.enrollVad.free()
    this.enrollVad = null
    const speechSeconds = this.enrollSpeechSamples / SAMPLE_RATE
    if (speechSeconds < this.enrollMin) {
      throw new Error(`净语音不足：${speechSeconds.toFixed(2)}s < ${this.enrollMin}s，请再说一句`)
    }
    // VoiceFilter.enroll 负责声纹提取与短注册阈值补偿（输入已是剥好静音的纯语音）
    return this.filter.enroll(concatFloat32(this.enrollChunks))
  }

  /** 喂入一段流式音频，返回自此完结的语音段及其门控判定。 */
  async accept(samples: Float32Array): Promise<GateSegmentEvent[]> {
    this.vad.acceptWaveform(samples)
    return this.drain()
  }

  /** 流结束：冲出尾部未完结的语音段并判定。 */
  async flush(): Promise<GateSegmentEvent[]> {
    this.vad.flush()
    return this.drain()
  }

  /** 复位 VAD 与段状态（注册声纹保留）。 */
  reset(): void {
    this.vad.reset()
  }

  dispose(): void {
    this.vad.free()
    this.enrollVad?.free()
    this.enrollVad = null
  }

  private collectEnrollSegments(): void {
    while (!this.enrollVad!.isEmpty()) {
      const segment = this.enrollVad!.front()
      this.enrollVad!.pop()
      this.enrollChunks.push(segment.samples)
      this.enrollSpeechSamples += segment.samples.length
    }
  }

  private async drain(): Promise<GateSegmentEvent[]> {
    const events: GateSegmentEvent[] = []
    while (!this.vad.isEmpty()) {
      const segment = this.vad.front()
      this.vad.pop()
      const judge = await this.filter.judge(segment.samples)
      events.push({
        type: 'segment',
        start: segment.start,
        durationSeconds: judge.durationSeconds,
        similarity: this.filter.enrolled ? judge.similarity : null,
        accepted: judge.accepted,
        samples: segment.samples,
      })
    }
    return events
  }
}

File: sdk/cpp-napi/src/full/speaker_gate.cpp
// 声纹门控完整版实现：sherpa-onnx cxx-api 的 VoiceActivityDetector（段式）切流，
// 声纹注册/判定复用 core 的 VoiceFilter；注册音频经独立 VAD 剥静音后再提声纹。
#include "audiotse/full/speaker_gate.hpp"

#include <algorithm>
#include <cstdio>
#include <optional>
#include <stdexcept>
#include <utility>

#include "sherpa-onnx/c-api/cxx-api.h"

namespace audiotse {
namespace {

constexpr float kVadBufferSizeSeconds = 30;

// sherpa 官方预编译库对大块单次喂入会丢失段边界（实测 5s 语音整段喂只检出 0.16s，
// WASM 版同样问题），统一切成 160ms 小块喂入，使行为与调用方块大小无关
// （与 web SDK 的 vad.ts 同一对策）。
class SegmentedVad {
public:
    static SegmentedVad Create(const SpeakerGateConfig &config) {
        sherpa_onnx::cxx::VadModelConfig vad_config;
        vad_config.silero_vad.model = config.vad_model;
        vad_config.silero_vad.threshold = config.vad_threshold;
        vad_config.silero_vad.min_silence_duration = config.min_silence_duration;
        vad_config.silero_vad.min_speech_duration = config.min_speech_duration;
        vad_config.silero_vad.max_speech_duration = config.max_speech_duration;
        vad_config.silero_vad.window_size = 512;
        vad_config.sample_rate = kSampleRate;
        vad_config.num_threads = config.vad_num_threads;
        vad_config.provider = "cpu";
        vad_config.debug = false;
        sherpa_onnx::cxx::VoiceActivityDetector vad =
            sherpa_onnx::cxx::VoiceActivityDetector::Create(vad_config, kVadBufferSizeSeconds);
        if (vad.Get() == nullptr) {
            throw std::runtime_error("failed to load vad model: " + config.vad_model);
        }
        return SegmentedVad(std::move(vad));
    }

    void AcceptWaveform(const float *samples, size_t n) {
        for (size_t off = 0; off < n; off += kFeedChunkSamples) {
            const size_t m = std::min(kFeedChunkSamples, n - off);
            vad_.AcceptWaveform(samples + off, static_cast<int32_t>(m));
        }
    }

    void Flush() { vad_.Flush(); }
    void Reset() { vad_.Reset(); }
    bool IsEmpty() const { return vad_.IsEmpty(); }
    sherpa_onnx::cxx::SpeechSegment Front() const { return vad_.Front(); }
    void Pop() { vad_.Pop(); }

private:
    explicit SegmentedVad(sherpa_onnx::cxx::VoiceActivityDetector &&vad) : vad_(std::move(vad)) {}

    static constexpr size_t kFeedChunkSamples = 2560;  // 160 ms @16k
    sherpa_onnx::cxx::VoiceActivityDetector vad_;
};

}  // namespace

struct SpeakerGate::Impl {
    SpeakerGateConfig config;
    SegmentedVad vad;
    VoiceFilter filter;
    /// 注册专用 VAD（与主流独立，剥静音用）；空 = 当前没有进行中的注册
    std::optional<SegmentedVad> enroll_vad;
    /// 注册累计的纯语音段
    std::vector<std::vector<float>> enroll_chunks;
    size_t enroll_speech_samples = 0;

    explicit Impl(const SpeakerGateConfig &c)
        : config(c), vad(SegmentedVad::Create(c)), filter(MakeFilterConfig(c)) {}

    static VoiceFilterConfig MakeFilterConfig(const SpeakerGateConfig &c) {
        VoiceFilterConfig fc;
        fc.speaker_model = c.speaker_model;
        fc.threshold = c.threshold;
        fc.short_enroll_threshold_factor = c.short_enroll_threshold_factor;
        fc.num_threads = c.embed_num_threads;
        return fc;
    }

    /// 把注册 VAD 中已完结的语音段搬进 enroll_chunks
    void CollectEnrollSegments() {
        while (!enroll_vad->IsEmpty()) {
            sherpa_onnx::cxx::SpeechSegment segment = enroll_vad->Front();
            enroll_vad->Pop();
            enroll_chunks.push_back(std::move(segment.samples));
            enroll_speech_samples += enroll_chunks.back().size();
        }
    }

    /// 取走主流 VAD 中已完结的语音段并逐段判定
    std::vector<GateSegmentEvent> Drain() {
        std::vector<GateSegmentEvent> events;
        while (!vad.IsEmpty()) {
            sherpa_onnx::cxx::SpeechSegment segment = vad.Front();
            vad.Pop();

            GateSegmentEvent event;
            event.start = segment.start;
            event.duration_seconds =
                static_cast<float>(segment.samples.size()) / static_cast<float>(kSampleRate);
            const JudgeResult judge = filter.Judge(segment.samples.data(), segment.samples.size());
            if (filter.Enrolled()) {
                event.similarity = judge.similarity;
                event.has_similarity = true;
            }
            event.accepted = judge.accepted;
            event.samples = std::move(segment.samples);
            events.push_back(std::move(event));
        }
        return events;
    }
};

// SegmentedVad::Create 在 VAD 模型加载失败时抛 std::runtime_error
SpeakerGate::SpeakerGate(const SpeakerGateConfig &config) : impl_(new Impl(config)) {}

SpeakerGate::~SpeakerGate() = default;

bool SpeakerGate::Enrolled() const { return impl_->filter.Enrolled(); }

float SpeakerGate::EffectiveThreshold() const { return impl_->filter.EffectiveThreshold(); }

EnrollResult SpeakerGate::Enroll(const float *samples, size_t n) {
    BeginEnroll();
    EnrollChunk(samples, n);
    return FinishEnroll();
}

void SpeakerGate::BeginEnroll() {
    impl_->enroll_vad = SegmentedVad::Create(impl_->config);
    impl_->enroll_chunks.clear();
    impl_->enroll_speech_samples = 0;
}

EnrollProgress SpeakerGate::EnrollChunk(const float *samples, size_t n) {
    if (!impl_->enroll_vad) BeginEnroll();
    impl_->enroll_vad->AcceptWaveform(samples, n);
    impl_->CollectEnrollSegments();
    const double speech_seconds = static_cast<double>(impl_->enroll_speech_samples) / kSampleRate;
    return EnrollProgress{speech_seconds,
                          speech_seconds >= impl_->config.enroll_target_speech_seconds};
}

EnrollResult SpeakerGate::FinishEnroll() {
    if (!impl_->enroll_vad) {
        throw std::runtime_error("注册未开始（先 BeginEnroll）");
    }
    impl_->enroll_vad->Flush();
    impl_->CollectEnrollSegments();
    impl_->enroll_vad.reset();

    const double speech_seconds = static_cast<double>(impl_->enroll_speech_samples) / kSampleRate;
    if (speech_seconds < impl_->config.enroll_min_speech_seconds) {
        char message[128];
        std::snprintf(message, sizeof(message), "净语音不足：%.2fs < %.2gs，请再说一句", speech_seconds,
                      impl_->config.enroll_min_speech_seconds);
        throw std::runtime_error(message);
    }
    // VoiceFilter.Enroll 负责声纹提取与短注册阈值补偿（输入已是剥好静音的纯语音）
    size_t total = 0;
    for (const std::vector<float> &chunk : impl_->enroll_chunks) total += chunk.size();
    std::vector<float> speech;
    speech.reserve(total);
    for (const std::vector<float> &chunk : impl_->enroll_chunks) {
        speech.insert(speech.end(), chunk.begin(), chunk.end());
    }
    return impl_->filter.Enroll(speech);
}

std::vector<GateSegmentEvent> SpeakerGate::AcceptWaveform(const float *samples, size_t n) {
    impl_->vad.AcceptWaveform(samples, n);
    return impl_->Drain();
}

std::vector<GateSegmentEvent> SpeakerGate::Flush() {
    impl_->vad.Flush();
    return impl_->Drain();
}

void SpeakerGate::Reset() { impl_->vad.Reset(); }

}  // namespace audiotse
